### PR TITLE
Refactor `backfilled` into specific behavior function arguments pt.1

### PR DESCRIPTION
**Pull request by MadLittleMods**
Refactor `backfilled` into specific behavior function arguments pt.1. This PR does not replace all `backfilled` instances. Just `persist_events_and_notify` and all downstream calls

Using keyword-only arguments (`*`) to reduce the mistakes from `backfilled` being left as a positional argument somewhere and being interpreted wrong by our new arguments.

Part of https://github.com/matrix-org/synapse/issues/11300

### Dev notes

 - `persist_events_and_notify` (added `inhibit_push_notifications`)
    - `persist_events` and `persist_event`
       - `_event_persist_queue`

`_event_persist_queue` -> 

 - `_persist_event_batch` (added `should_calculate_state_and_forward_extrems`)
    - `_persist_events_and_state_updates` (added `use_negative_stream_ordering`)
       - `_persist_events_txn`
          - `_update_room_depths_txn` (added `update_room_forward_stream_ordering`)
          - `_update_metadata_tables_txn`
             - `_store_room_members_txn` (added `inhibit_local_membership_updates`)

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [ ] ~~Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)~~
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/replication/http/federation.py [hunk 1]
```diff
@@ -69,14 +69,37 @@ def __init__(self, hs: "HomeServer"):
...
-    async def _serialize_payload(store, room_id, event_and_contexts, backfilled):
+    async def _serialize_payload(
+        store,
+        room_id,
+        event_and_contexts,
+        inhibit_push_notifications,
+        should_calculate_state_and_forward_extrems,
+        use_negative_stream_ordering,
+        inhibit_local_membership_updates,
+        update_room_forward_stream_ordering,
```
Comment (MadLittleMods) [inline synapse/replication/http/federation.py:80]: Is there a way we can define the arguments here like the following? I can't get it to work even without the keyword-only marker.

```py
async def _serialize_payload(
    store,
    room_id,
    event_and_contexts,
    *,
    inhibit_push_notifications: bool = False,
    should_calculate_state_and_forward_extrems: bool = True,
    use_negative_stream_ordering: bool = False,
    inhibit_local_membership_updates: bool = False,
    update_room_forward_stream_ordering: bool = True,
):
```
    
When I do this, I get the following error:

```
synapse/replication/http/federation.py:72: error: Signature of "_serialize_payload" incompatible with supertype "ReplicationEndpoint"  [override]
```

And I can't figure out how to adjust the abstract method to be more flexible, https://github.com/matrix-org/synapse/blob/7ffddd819c2e0e7edff141e987372205894084b6/synapse/replication/http/_base.py#L131-L132

File: synapse/storage/persist_events.py [hunk 2]
```diff
@@ -35,6 +35,7 @@
 )
 
 import attr
+from mypy_extensions import DefaultNamedArg
```
Comment (MadLittleMods) [inline synapse/storage/persist_events.py:38]: Is there a better way to tackle this?

The build is failing because I used it `builtins.ModuleNotFoundError: No module named 'mypy_extensions'`, https://github.com/matrix-org/synapse/runs/4261552290?check_suite_focus=true#step:6:739

File: synapse/handlers/federation_event.py [hunk 3]
```diff
@@ -1809,7 +1809,24 @@ async def _run_push_actions_and_persist_event(
...
+                should_calculate_state_and_forward_extrems=not backfilled,
+                # Backfilled events get a negative stream ordering so they don't
+                # come down incremental `/sync`
+                use_negative_stream_ordering=backfilled,
+                # Backfilled events do not affect the current local state
+                inhibit_local_membership_updates=backfilled,
+                # Backfilled events have negative stream_ordering and happened
+                # in the past so we know that we don't need to update the
+                # stream_ordering tip for the room.
+                update_room_forward_stream_ordering=not backfilled,
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py:1829]: In a future PR, I'll extend the behavior arguments further upstream in place of `backfilled`.

This PR does not replace all `backfilled` instances. Just `persist_events_and_notify` and all downstream calls

File: synapse/handlers/federation_event.py [hunk 4]
```diff
@@ -1859,15 +1904,19 @@ async def persist_events_and_notify(
...
+                update_room_forward_stream_ordering=update_room_forward_stream_ordering,
             )
 
             if self._ephemeral_messages_enabled:
                 for event in events:
                     # If there's an expiry timestamp on the event, schedule its expiry.
                     self._message_handler.maybe_schedule_expiry(event)
 
-            if not backfilled:  # Never notify for backfilled events
+            if not inhibit_push_notifications:  # Never notify for backfilled events
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py:1919]: Behavioral argument replacement: `inhibit_push_notifications`

File: synapse/storage/persist_events.py [hunk 5]
```diff
@@ -448,7 +590,7 @@ async def _persist_event_batch(
             # device lists as stale.
             potentially_left_users: Set[str] = set()
 
-            if not backfilled:
+            if should_calculate_state_and_forward_extrems:
```
Comment (MadLittleMods) [inline synapse/storage/persist_events.py:593]: Behavioral argument replacement: `should_calculate_state_and_forward_extrems`

File: synapse/storage/databases/main/events.py [hunk 6]
```diff
@@ -159,7 +175,7 @@ async def _persist_events_and_state_updates(
         #
         # Note: Multiple instances of this function cannot be in flight at
         # the same time for the same room.
-        if backfilled:
+        if use_negative_stream_ordering:
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/events.py:178]: Behavior argument replacement: `use_negative_stream_ordering`

File: synapse/storage/databases/main/events.py [hunk 7]
```diff
@@ -1682,7 +1740,7 @@ def non_null_str_or_none(val: Any) -> Optional[str]:
             # band membership", like a remote invite or a rejection of a remote invite.
             if (
                 self.is_mine_id(event.state_key)
-                and not backfilled
+                and not inhibit_local_membership_updates
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/events.py:1743]: Behavioral argument replacement: `inhibit_local_membership_updates`

File: synapse/storage/databases/main/events.py [hunk 8]
```diff
@@ -176,13 +192,15 @@ async def _persist_events_and_state_updates(
...
+                inhibit_local_membership_updates=inhibit_local_membership_updates,
+                update_room_forward_stream_ordering=update_room_forward_stream_ordering,
                 state_delta_for_room=state_delta_for_room,
                 new_forward_extremeties=new_forward_extremeties,
             )
             persist_event_counter.inc(len(events_and_contexts))
 
-            if not backfilled:
+            # TODO: test that this actuall works
+            if stream < 0:
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/events.py:203]: TODO: Test that this actually works

File: synapse/storage/databases/main/events.py [hunk 9]
```diff
@@ -1200,21 +1232,27 @@ def _update_room_depths_txn(
...
+                events because backfilled events have negative stream_ordering
+                and happened in the past so we know that we don't need to
+                update the stream_ordering tip for the room.
         """
         depth_updates: Dict[str, int] = {}
         for event, context in events_and_contexts:
             # Remove the any existing cache entries for the event_ids
             txn.call_after(self.store._invalidate_get_event_cache, event.event_id)
-            if not backfilled:
+            if update_room_forward_stream_ordering:
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/events.py:1255]: Behavioral argument replacment: `update_room_forward_stream_ordering`

File: synapse/handlers/federation_event.py [hunk 10]
```diff
@@ -1821,7 +1838,12 @@ async def persist_events_and_notify(
         self,
         room_id: str,
         event_and_contexts: Sequence[Tuple[EventBase, EventContext]],
-        backfilled: bool = False,
+        *,
+        inhibit_push_notifications: bool = False,
+        should_calculate_state_and_forward_extrems: bool = True,
+        use_negative_stream_ordering: bool = False,
+        inhibit_local_membership_updates: bool = False,
+        update_room_forward_stream_ordering: bool = True,
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py:1846]: @richvdh How does this look? Is this on the right track?

File: synapse/replication/http/federation.py (hunk 1 above)
Comment (richvdh) [reply synapse/replication/http/federation.py:80]: I'm not entirely sure why, but it's the `:bool` type annotations that are upsetting it rather than the `*` or the default values. If I remove the `:bool`s, it's fine.

File: synapse/handlers/federation_event.py (hunk 3 above)
Comment (richvdh) [inline synapse/handlers/federation_event.py:1829]: shouldn't `update_room_forward_stream_ordering` be implied by `use_negative_stream_ordering` ?

File: synapse/handlers/federation_event.py [hunk 11]
```diff
@@ -1809,7 +1809,24 @@ async def _run_push_actions_and_persist_event(
...
+                # We don't need to calculate the state for backfilled events and
+                # there is no need to update the forward extrems because we
+                # already know this event happened in the past if it was
+                # backfilled.
+                should_calculate_state_and_forward_extrems=not backfilled,
+                # Backfilled events get a negative stream ordering so they don't
+                # come down incremental `/sync`
+                use_negative_stream_ordering=backfilled,
+                # Backfilled events do not affect the current local state
+                inhibit_local_membership_updates=backfilled,
```
Comment (richvdh) [inline synapse/handlers/federation_event.py:1825]: would it be sensible to have this implied by `should_calculate_state_and_forward_extrems` ?

File: synapse/handlers/federation_event.py [hunk 12]
```diff
@@ -1809,7 +1809,24 @@ async def _run_push_actions_and_persist_event(
...
-                event.room_id, [(event, context)], backfilled=backfilled
+                event.room_id,
+                [(event, context)],
+                # We should not send notifications about backfilled events.
+                inhibit_push_notifications=backfilled,
+                # We don't need to calculate the state for backfilled events and
+                # there is no need to update the forward extrems because we
+                # already know this event happened in the past if it was
+                # backfilled.
+                should_calculate_state_and_forward_extrems=not backfilled,
```
Comment (richvdh) [inline synapse/handlers/federation_event.py:1820]: the `should_` prefix seems a bit redundant here.

File: synapse/handlers/federation_event.py [hunk 13]
```diff
@@ -1850,7 +1895,7 @@ async def persist_events_and_notify(
                     store=self._store,
                     room_id=room_id,
                     event_and_contexts=batch,
-                    backfilled=backfilled,
+                    inhibit_push_notifications=inhibit_push_notifications,
```
Comment (richvdh) [inline synapse/handlers/federation_event.py:1898]: doesn't this need to pass through *all* the keyword args?

File: synapse/storage/databases/main/events.py (hunk 6 above)
Comment (richvdh) [reply synapse/storage/databases/main/events.py:178]: > Behavior argument replacement: `use_negative_stream_ordering`

sorry, what does this mean?

**Review by richvdh — CHANGES_REQUESTED** [review]
Thanks for putting this up, @MadLittleMods!

I haven't read all the way through the diff, because it's a bit hard to follow. Could I ask you to break it down a bit - let's start somewhere nearer the bottom of the stack: say `_persist_events_and_state_updates`. I think that will make it much easier to convince ourselves that the changes are correct.

Generally it seems like the right *sort* of thing, but I fear we might be taking it to extremes.  If there are ways of grouping some of these flags together into coherent bits of functionality, I think that would be better. For example - does it ever make sense to set `use_negative_stream_ordering=True` and `update_room_forward_stream_ordering=True` together (or both `False`)? If not, let's get rid of `update_room_forward_stream_ordering` - otherwise it's just a nightmare to use the function.

File: synapse/storage/databases/main/events.py (hunk 6 above)
Comment (MadLittleMods) [reply synapse/storage/databases/main/events.py:178]: It's just the root place that I originally replaced and defines the behavior

File: synapse/handlers/federation_event.py (hunk 11 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py:1825]: I understand the leap that updating the state in the room also encompasses member state events.

My first impression was that it wasn't very obvious but it makes sense ⏩

File: synapse/handlers/federation_event.py (hunk 13 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py:1898]: 👍 I think so. Should match `_serialize_payload`.

It's a loose connection and I was having trouble typing it correctly, https://github.com/matrix-org/synapse/pull/11396#discussion_r752929555, to have the linter yell at me for this.

**Comment by MadLittleMods** [comment]
Thanks for the review @richvdh, great points! I've made a smaller PoC in https://github.com/matrix-org/synapse/pull/11417
